globus_action_provider_tools/data_types.py

```diff
@@ -84,31 +86,93 @@ class ActionProviderDescription:
+        regex="^P(?!$)(\d+(?:\.\d+)?Y)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?W)?(\d+(?:\.\d+)?D)?(T(?=\d)(\d+(?:\.\d+)?H)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?S)?)?$",
+    )
+    monitor_by: Set[str] = Field(
+        default_factory=list,
+        description="A list of principal URNs containing identities which are allowed to monitor the progress of the action using the status and log operations. When not provided, defaults to the user that initiated the action.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    manage_by: Set[str] = Field(
+        default_factory=list,
+        description="A list of principal URNs containing identities which are allowed to manage the progress of the action using the cancel and release operations. When not provided, defaults to the user that initiated the action.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    allowed_clients: List[str] = Field(
+        default_factory=list, regex="^(public|globus|creator|.$)$"
+    )
+
+
+class ActionStatus(BaseModel):
+    status: ActionStatusValue = Field(
+        ..., description="The current state of the Action"
+    )
+    creator_id: str = Field(
+        ...,
+        description="A URN representation of an Identity in Globus either of a user from Globus Auth or a group from Globus Groups.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    action_id: str = Field(
+        default_factory=shortish_id, description="The id of the Action itself"
+    )
+    start_time: str = Field(default_factory=now_isoformat)
+    label: Optional[str] = Field(
+        None,
+        description="A short human presentable description of the Action requested",
+        min_length=1,
+        max_length=64,
+    )
+    monitor_by: Set[str] = Field(
+        default_factory=list,
+        description="A list of principal URNs containing identities which are allowed to monitor the progress of the action using the /status and /log operations. When not provided, defaults to the user that initiated the action.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    manage_by: Set[str] = Field(
+        default_factory=list,
+        description="A list of principal URNs containing identities which are allowed to manage the progress of the action using the cancel and release operations. When not provided, defaults to the user that initiated the action.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    completion_time: Optional[datetime.datetime] = Field(
```

Comment:
Does using type `datetime.datetime` get us any iso representation validation via pydantic? Should there be a regex of some sort if not (though perhaps that's not good since it is not a `str` type)

Reply:
Yup, datetimes will be parsed as iso8601. I think pydantic supports a few different formats when converting to a datetime, and iso8601 is just one of them. I hope that the description of the field is enough to put a user on the right path.